docs/modern-chialisp.md

```diff
@@ -0,0 +1,1438 @@
+itself if we just want to pass it on whole or return it.
+
+You can also use this to simulate Maybe or Options type in a convenient way.
+
+```haskell
+> f p = fromMaybe 0 $ (\(Pt x y) -> x + y) <$> p
+> f (Just (Pt 3 5))
+8
+> f Nothing
+0
+```
+
+In chialisp:
+
+
+```chialisp
+(mod (p) 
+  (include *standard-cl-23*) 
+      
+  (defun F ((@ p (x y))) (if p (+ x y) 0)) 
+      
+  (F p)
+  )
+```
+```shell
+$ brun maybe.clvm '((3 5))'
+8
+$ brun maybe.clvm '(())'
+0
+```
+
+You can also do this with a higher order functions similarly.  In this
+implementation, maybe is always a list so that (Just 0) is distinct from
+Nothing.  This is longer but more general.  If used to construct fully
+generic algorithms, it might be needed.
+
+```chialisp
+(mod (p)
+    
+  (include *standard-cl-23*)
+    
+  (defun fromMaybe (v (@ m (content))) (if m content v))
+  (defun mapMaybe (f (@ m (content))) (if m (list (a f (list content))) ()))
+    
+  (fromMaybe 0 (mapMaybe (lambda ((@ pt (x y))) (+ x y)) p))
+  )
+```
+```shell
+$ brun maybe.clvm '(((3 5)))' # Note: ((3 5)) is like (Some (Pt 3 5))
+8
+$ brun maybe.clvm '(())'
+()
+```
+
+### optional types
+
+There is an optional type system described [here](/docs/types-tutorial.md)
+
+### Function closures
+
+Chialisp allows function names to be treated as closures when used as values.
+The resulting function is callable with the 'a' operator in the way other
+foreign code would be when passed into your program or as a result of a lambda
+expression.
+
+```chialisp
+(mod (Z)
+   
+  (include *standard-cl-23*)
+    
+  (defun add-one (X) (+ X 1))
+    
+  (defun map (F L)
+    (if L
+      (c (a F (list (f L))) (map F (r L)))
+      ()
+      )
+    )
+    
+  (map add-one Z)
+  )
+```
+```shell
+$ brun map.clvm '((1 2 3))'
+(a 3 4)
+```
+
+### lambda forms
+
+Chialisp supports 'lambda' in a similar way to scheme.  The resulting value is
+callable with the 'a' operator, has access to the program's functions and constants
+and is in general safe to use for higher-order functions.
+
+The above program could be re-stated as:
+
+    (mod (Z)
+   
+      (include *standard-cl-23*)
+    
+      (defun map (F L)
+        (if L
+          (c (a F (list (f L))) (map F (r L)))
+          ()
+          )
+        )
+    
+      (map (lambda (Y) (+ Y 1)) Z)
+      )
+
+### embed and compile include forms
+
+Chialisp programs often need to know the hashes or contents of other programs
+and other data and it can be inconvenient to translate it into chialisp source
+code (especially if, during development it changes).
+
+Chialisp now gives the ability to embed foreign data, including compiling
+programs and embedding their compiled representation.  Since chialisp programs
+contain identification about how they should be compiled, it's possible to
+include programs from different versions of the language accurately.
+
+```chialisp
+(mod (Z)
+    
+  (include *standard-cl-23*)
+    
+  (embed-file hello-data bin "hello.txt")
```

Comment:
how about compile-include forms? is there a different syntax for that?

Reply:
Hasn't been merged yet, getting there.

Reply:
Thanks, will add.